Fix changes in option autoload not reflecting in object cache

kkmuffme opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/51352

kkmuffme committed:
Fix changes in option autoload not reflecting in object cache

https://core.trac.wordpress.org/ticket/51352

felixarntz commented on src/wp-includes/option.php:
```diff
@@ -545,14 +545,30 @@ function update_option( $option, $value, $autoload = null ) {
 	}
 
-	if ( ! wp_installing() ) {
+	if ( ! wp_installing() && null === $autoload ) {
 		$alloptions = wp_load_alloptions( true );
 		if ( isset( $alloptions[ $option ] ) ) {
 			$alloptions[ $option ] = $serialized_value;
 			wp_cache_set( 'alloptions', $alloptions, 'options' );
 		} else {
 			wp_cache_set( $option, $serialized_value, 'options' );
 		}
+	} elseif ( ! wp_installing() && 'yes' === $update_args['autoload'] ) {
+		// delete it before we force the cache
+		wp_cache_delete( $option, 'options' );
+
+		$alloptions = wp_load_alloptions( true );
+		$alloptions[ $option ] = $serialized_value;
+		wp_cache_set( 'alloptions', $alloptions, 'options' );
+	} elseif ( ! wp_installing() ) {
+		// not autoloaded
+		$alloptions = wp_load_alloptions( true );
+		if ( isset( $alloptions[ $option ] ) ) {
+			unset( $alloptions[ $option ] );
+			wp_cache_set( 'alloptions', $alloptions, 'options' );
+		}
+
+		wp_cache_set( $option, $serialized_value, 'options' );
 	}
```
It would be a bit simpler to not have the `! wp_installing()` check in every condition, and instead use nested conditions.
```suggestion
	if ( ! wp_installing() ) {
		if ( ! isset( $update_args['autoload'] ) ) {
			// Update the cached value based on where it is currently cached.
			$alloptions = wp_load_alloptions( true );
			if ( isset( $alloptions[ $option ] ) ) {
				$alloptions[ $option ] = $serialized_value;
				wp_cache_set( 'alloptions', $alloptions, 'options' );
			} else {
				wp_cache_set( $option, $serialized_value, 'options' );
			}
		} elseif ( 'yes' === $update_args['autoload'] ) {
			// Delete the individual cache, then set in alloptions cache.
			wp_cache_delete( $option, 'options' );

			$alloptions = wp_load_alloptions( true );
			$alloptions[ $option ] = $serialized_value;
			wp_cache_set( 'alloptions', $alloptions, 'options' );
		} else {
			// Delete the alloptions cache, then set the individual cache.
			$alloptions = wp_load_alloptions( true );
			if ( isset( $alloptions[ $option ] ) ) {
				unset( $alloptions[ $option ] );
				wp_cache_set( 'alloptions', $alloptions, 'options' );
			}

			wp_cache_set( $option, $serialized_value, 'options' );
		}
	}
```

kkmuffme replied:
Definitely, I PRed it like that to show that the existing code is unchanged. I'll committed your suggestion now

kkmuffme replied:
There are tests on the original PR https://github.com/WordPress/wordpress-develop/pull/541/commits/f81720085010a8c87a8e5116c190bbd7c467cfb2 - can I just cherry pick them to this PR? (not sure if that works with Github forks)

felixarntz replied:
@kkmuffme Yes, please you can simply copy them over here. One additional feedback on the tests: It would be great if you could split them apart into separate test methods, with only one thing being tested per test method. Typically this means one assertion per test.

kkmuffme replied:
@felixarntz the tests were non-class-scoped, which is why I copied them and adjusted.

Sorry I can't do any further changes here atm, since I'm out of time and not available until the 12th

felixarntz reviewed (commented):
@kkmuffme The PR looks great to me, just one comment below to simplify the if clauses.

The other thing is that we need test coverage for this change. Would you be able to add that to your PR?

kkmuffme committed:
Update src/wp-includes/option.php

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

kkmuffme committed:
cherry-picked tests

Taken from patch PR https://github.com/WordPress/wordpress-develop/commit/f81720085010a8c87a8e5116c190bbd7c467cfb2 props @pentatonicfunk

kkmuffme committed:
Merge branch 'trunk' into update_option-autoload-change-inconsistent-cache-leak

Commit:
Simplify complex test by splitting it and using only the crucial assertion.

felixarntz reviewed (approved):
Thank you @kkmuffme!

Based on your reply in https://github.com/WordPress/wordpress-develop/pull/4494#discussion_r1348865976, I have revised the tests a bit to simplify them. This looks good to go now.

felixarntz commented:
Committed in https://core.trac.wordpress.org/changeset/56796
